### PR TITLE
[Clang] Fix looking for immediate calls in default arguments.

### LINKED ISSUE #80630: [Clang] nontrivial use of std::source_location::current() in a default argument fails
The following
```
#include <source_location>

#define BOOST_CURRENT_LOCATION_PTR \
    []( char const* fn ) \
    { \
        static constexpr std::source_location loc = std::source_location::current(); \
        return &loc; \
    }( std::source_location::current().function_name() )

auto f( std::source_location const* loc = BOOST_CURRENT_LOCATION_PTR )
{
    return loc;
}

auto g()
{
    return f();
}
```
fails (for Clang trunk and Clang 17, with both libstdc++ and libc++) with
```
error: cannot take address of consteval function 'current' outside of an immediate invocation
```
(https://godbolt.org/z/jqqxq6njh)

GCC, MSVC accept.

The example is simplified to the point of doing not much at all, but it started out as something doing useful things. :-)

### PR DESCRIPTION
**Pull request by cor3ntin**
Due to improper use of RecursiveASTVisitor.

Fixes #80630

### CODE REVIEW

File: clang/lib/Sema/SemaExpr.cpp [hunk 1]
```diff
@@ -6222,9 +6222,9 @@ struct ImmediateCallVisitor : public RecursiveASTVisitor<ImmediateCallVisitor> {
 
   // Blocks don't support default parameters, and, as for lambdas,
   // we don't consider their body a subexpression.
-  bool VisitBlockDecl(BlockDecl *B) { return false; }
+  bool VisitBlockDecl(BlockDecl *B) { return true; }
 
-  bool VisitCompoundStmt(CompoundStmt *B) { return false; }
+  bool VisitCompoundStmt(CompoundStmt *B) { return true; }
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaExpr.cpp:6227]: I think we can just remove these entirely; the default implementation already returns true.

File: clang/lib/Sema/SemaExpr.cpp [hunk 2]
```diff
@@ -6198,7 +6198,7 @@ struct ImmediateCallVisitor : public RecursiveASTVisitor<ImmediateCallVisitor> {
   bool VisitCallExpr(CallExpr *E) {
     if (const FunctionDecl *FD = E->getDirectCallee())
       HasImmediateCalls |= FD->isImmediateFunction();
-    return RecursiveASTVisitor<ImmediateCallVisitor>::VisitStmt(E);
+    return RecursiveASTVisitor<ImmediateCallVisitor>::VisitCallExpr(E);
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaExpr.cpp:6201]: This is a noop, right? If so, I'd drop the changes here but make an NFC change for this and the one in `VisitSourceLocExpr()` at the same time.

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Comment by zyn0217** [comment]
Does this fix https://github.com/llvm/llvm-project/issues/67134 as well?

**Comment by cor3ntin** [comment]
> Does this fix #67134 as well?

I don't think so, no

**Pull request merged**